### PR TITLE
feature/gitleaks-scanner (closes #135)

### LINKED ISSUE #135: Add a new a git credential scanner (gitleaks)
## New Scanner implementation request

**Is your feature request related to a problem? Please describe.**
As a security tester i would like to continuously check a list of git repositories in my project and check if they contain any credentials. This is a common security problem which must be avoided. But even the best developers can make mistakes 🤷🏻‍♂️. Finding such problems early can prevent worse.

**Describe the solution you'd like**
There are alot of different git credential scanner services and tools. The gitleaks (SAST) scanner seems to be a good solution which can be easily integrated with the SCBv2 Project: https://github.com/zricethezav/gitleaks

**Describe alternatives you've considered**

**Additional context**

## Steps to implement a new scanner

- [ ] Create a new folder with the name of the [scanner here](https://github.com/secureCodeBox/secureCodeBox-v2/tree/master/scanners)
- [ ] Add a README.md and give a brief overview of the scanner and its configuration options.
- [ ] Implement a new scanner specific scan-type.yaml
- [ ] Implement a new scanner specific parse-definition.yaml
- [ ] Add (optional) some cascading-rules.yaml
- [ ] Add (optional) a Dockerfile for the scanner if there is no existing one publicly available on dockerHub
- [ ] Use the [parser-SDK](https://github.com/secureCodeBox/secureCodeBox-v2/tree/master/parser-sdk) to implement a new findings parser (currently based on NodeJS)
- [ ] Add unit tests with at minimum 80% test coverage

### PR DESCRIPTION
**Pull request by paulschmelzer**
**Adds a new scanner (closes #135).**

Gitleaks is a git repository scanner for detecting secrets in files.

### CODE REVIEW

File: scanners/gitleaks/parser/parser.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
+HIGH_TAGS = ['JWT', 'Artifactory', 'AWS', 'PrivateKey'];
+MEDIUM_TAGS = ['Hash', 'Facebook', 'Twitter', 'Github', 'LinkedIn', 'Slack', 'Google', 'Heroku',
+  'Mailchimp', 'Mailgun', 'Paypal', 'Picatic', 'Teams', 'Jenkins', 'Stripe', 'Square', 'Twilio'];
+LOW_TAGS = ['Generic', 'FileName', 'FileExtension'];
```
Comment (J12934) [inline scanners/gitleaks/parser/parser.js:4]: These should be defined using const or let.

```suggestion
const HIGH_TAGS = ['JWT', 'Artifactory', 'AWS', 'PrivateKey'];
const MEDIUM_TAGS = ['Hash', 'Facebook', 'Twitter', 'Github', 'LinkedIn', 'Slack', 'Google', 'Heroku',
  'Mailchimp', 'Mailgun', 'Paypal', 'Picatic', 'Teams', 'Jenkins', 'Stripe', 'Square', 'Twilio'];
const LOW_TAGS = ['Generic', 'FileName', 'FileExtension'];
```

Comment (J12934) [reply scanners/gitleaks/parser/parser.js:4]: LOW_TAGS also seems unused

File: scanners/gitleaks/helm2.Chart.yaml [hunk 2]
```diff
@@ -0,0 +1,23 @@
+apiVersion: v2
```
Comment (rseedorff) [inline scanners/gitleaks/helm2.Chart.yaml:1]: Must be `apiVersion: v1`

File: scanners/gitleaks/values.yaml [hunk 3]
```diff
@@ -0,0 +1,45 @@
+parserImage:
+  # parserImage.tag - defaults to the charts version
+  # parserImage.repository -- Parser image repository
+  repository: paulschmelzer/gitleaks-parser
```
Comment (rseedorff) [inline scanners/gitleaks/values.yaml:4]: Should be changed to repository: `docker.io/securecodebox/gitleaks-parser` ;-)

File: scanners/gitleaks/values.yaml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  # parserImage.tag -- Parser image tag
+  # @default -- defaults to the charts version
+  tag: null
+
+image:
+  repository: docker.io/paulschmelzer/gitleaks-wrapper
```
Comment (rseedorff) [inline scanners/gitleaks/values.yaml:10]: Should be changed to repository: `docker.io/securecodebox/gitleaks-wrapper` ;-)

File: scanners/gitleaks/parser/parser.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    } else if (containsTag(finding.tags, MEDIUM_TAGS)) {
+      severity = 'MEDIUM'
+    }
+
+    return {
+      name: finding.rule,
+      description: 'The name of the rule which triggered the finding: ' + finding.rule,
+      osi_layer: 'APPLICATION',
+      severity: severity,
+      category: 'potential secret',
```
Comment (J12934) [inline scanners/gitleaks/parser/parser.js:23]: Our other categories are all capitalized.

```suggestion
      category: 'Potential Secret',
```

File: scanners/gitleaks/helm2.Chart.yaml [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+keywords:
+  - security
+  - gitleaks
+  - scanner
+  - secureCodeBox
+home: https://docs.securecodebox.io/docs/scanners/gitleaks
+icon: https://docs.securecodebox.io/img/integrationIcons/gitleakslogo.png
+sources:
+  - https://github.com/secureCodeBox/secureCodeBox-v2
```
Comment (J12934) [inline scanners/gitleaks/helm2.Chart.yaml:19]: Points to the old repo ;)

Comment (J12934) [inline scanners/gitleaks/helm2.Chart.yaml:19]: Also points to old repo

**Review by rseedorff — CHANGES_REQUESTED** [review]
Hi @paulschmelzer,
thx a lot for your contribution 🤗 looks great so far! Keep on going... If you are fast you PR will make it into our next major release 2.0.0 ... ;-)
Can you add an update to the `/secureCodeBox/.github/workflows/ci.yaml` file and add your new scanner implementation into the ci pipeline? I guess @J12934 will assist you to create a new corresponding docker repo.

File: scanners/gitleaks/README.md.gotmpl [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+It is not an easy task to classify the severity of the scans because we can't tell for sure if the finding is e.g. a real
+or a testing password. Another issue is that the rate of false positives for generic rules can be very high. Therefore, 
+we tried to classify the severity of the finding by looking at the accuracy of the rule which detected it. Rules for AWS 
+secrets or Artifactory tokens are very precise, so they get a high severity. Generic rules on the other hand get a low 
+severity because the often produce false positives. 
+
+**Please keep in mind that findings with a low severity can be actually
+very critical.**
+
+#### Examples
```
Comment (J12934) [inline scanners/gitleaks/README.md.gotmpl:61]: We use a somewhat complicated mechanism in the other scanner folder to generate the examples from standalone files.

Basically it should have a folder structure like the following.

```
gitleaks/
├─ examples/
│  ├─ multi-juicer/
│  │  ├─ Scan.yaml
│  │  ├─ README.md
│  ├─ private-repo/
│  │  ├─ Scan.yaml
│  │  ├─ README.md
├─ Chart.yaml
├─ Other files and folders
├─ ...

```

File: scanners/gitleaks/README.md.gotmpl [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+**Do not** override the option `--report-format` or `--report`. It is already configured for automatic findings parsing.
+
+#### Ruleset
+At this point we provide three rulesets which you can pass to the `--config` oprtion:
+
+- `/home/config_all.toml`: Includes every rule.
+- `/home/config_filenames_only.toml`: Gitleaks scans only file names and extensions.
+- `/home/config_no_generics.toml`: No generic rules like searching for the word *password*. With this option you won't
+find something like **password = Ej2ifDk2jfeo2** but it will reduce resulting false positives.
```
Comment (J12934) [inline scanners/gitleaks/README.md.gotmpl:44]: That's really nice to provide default configs 👍

**Review by J12934 — CHANGES_REQUESTED** [review]

**Review by J12934 — APPROVED** [review]
